src/libavrdude.i

```diff
@@ -0,0 +1,561 @@
+  if ($input != Py_None && !PyCallable_Check($input)) {
+      PyErr_SetString(PyExc_TypeError, "Need a callable object!");
+      return NULL;
+  }
+  $1 = $input;
+}
+void set_msg_callback(PyObject *PyFunc);
+void set_progress_callback(PyObject *PyFunc);
+
+// These things are read from config file(s), and must be considered
+// read-only by any program. Most internals are only relevant for
+// libavrdude itself, so only map things to the Python level that are
+// needed there.
+
+%typemap(out) unsigned char signature[3] {
+  $result = PyBytes_FromStringAndSize((const char *)$1, 3);
+ }
+
+%immutable;
+typedef struct avrpart {
+  const char  * desc;               /* long part name */
+  const char  * id;                 /* short part name */
+  LISTID        variants;           /* String with variant name and chip properties */
+  const char  * parent_id;          /* Used by developer options */
+  const char  * family_id;          /* family id in the SIB (avr8x) */
+  int           prog_modes;         /* Programming interfaces, see #define PM_... */
+  unsigned char signature[3];       /* expected value of signature bytes */
+  unsigned short usbpid;            /* USB DFU product ID (0 = none) */
+  LISTID        mem;            /* avr memory definitions */
+  LISTID        mem_alias;      /* memory alias definitions */
+  const char  * config_file;    /* config file where defined */
+  int           lineno;         /* config file line number */
+} AVRPART;
+%mutable;
+
+typedef unsigned int memtype_t;
+typedef struct avrmem {
+  %immutable;
+  const char *desc;           /* memory description ("flash", "eeprom", etc) */
+  memtype_t type;             /* internally used type, cannot be set in conf files */
+  bool paged;                  /* 16-bit page addressed, e.g., ATmega flash but not EEPROM */
+  int size;                   /* total memory size in bytes */
+  int page_size;              /* size of memory page (if page addressed) */
+  int num_pages;              /* number of pages (if page addressed) */
+  int initval;                /* factory setting of fuses and lock bits */
+  int bitmask;                /* bits used in fuses and lock bits */
+  %mutable;
+  unsigned char * buf;        /* pointer to memory buffer */
+} AVRMEM;
+
+%extend avrmem {
+  PyObject *get(unsigned int len, unsigned int offset = 0) {
+    if (offset > (unsigned)$self->size)
+      return Py_None;
+    if (offset + len > (unsigned)$self->size)
+      len = $self->size - offset;
+    return PyBytes_FromStringAndSize((char *)($self->buf + offset), len);
+  }
+}
+
+int avr_initmem(const AVRPART *p);
+
+%extend avrmem {
+  %typemap(in) (unsigned char *in, unsigned int len) {
+    Py_ssize_t len;
+    PyBytes_AsStringAndSize($input, (char **)&$1, &len);
+    $2 = len;
+  }
+  int put(unsigned char *in, unsigned int len, unsigned int offset = 0) {
+    if ($self->buf == NULL)
+      // missing avr_initmem()?
+      return 0;
+    if (offset > (unsigned)$self->size)
```

Comment:
This is not wrong (b/c the lines below heal it) but should it not be >= instead of >? Same above for get

Reply:
Good catch, fixed.